File: Legacy/Back-end/CatFactory.UI.API/CatFactory.UI.API/Models/Responses.cs
using System;
using System.Collections.Generic;
using System.Net;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;

namespace CatFactory.UI.API.Models
{
    public interface IResponse
    {
        string Message { get; set; }

        bool DidError { get; set; }

        string ErrorMessage { get; set; }
    }

    public interface IListResponse<TModel> : IResponse
    {
        IEnumerable<TModel> Model { get; set; }
    }

    public interface ISingleResponse<TModel> : IResponse
    {
        TModel Model { get; set; }
    }

    public class ImportDatabaseResponse : IResponse
    {
        public string Message { get; set; }

        public bool DidError { get; set; }

        public string ErrorMessage { get; set; }
    }

    public class ListResponse<TModel> : IListResponse<TModel>
    {
        public string Message { get; set; }

        public bool DidError { get; set; }

        public string ErrorMessage { get; set; }

        public IEnumerable<TModel> Model { get; set; }
    }

    public class SingleResponse<TModel> : ISingleResponse<TModel>
    {
        public string Message { get; set; }

        public bool DidError { get; set; }

        public string ErrorMessage { get; set; }

        public TModel Model { get; set; }
    }

    public static class ResponsesExtensions
    {
        public static void SetError(this IResponse response, Exception ex, ILogger logger)
        {
            logger?.LogCritical(ex.ToString());

            response.DidError = true;
            response.ErrorMessage = ex.Message;
        }

        public static IActionResult ToHttpResponse(this IResponse response)
        {
            var status = HttpStatusCode.OK;

            if (response.DidError)
                status = HttpStatusCode.InternalServerError;

            return new ObjectResult(response)
            {
                StatusCode = (int)status
            };
        }
    }

    public class EditDescription
    {
    }

    public class ImportedDatabase
    {
        public string Name { get; set; }

        public string Dbms { get; set; } = "SQL Server";

        public int TablesCount { get; set; }

        public int ViewsCount { get; set; }

        public string Details { get; } = "Details";
    }
}

File: Frontend/Legacy/cat-factory-ui/src/app/responses.ts
export class ListResponse<TModel> {
    public message: string;
    public didError: boolean;
    public errorMessage: string;
    public model: TModel[];
}

export class SingleResponse<TModel> {
    public message: string;
    public didError: boolean;
    public errorMessage: string;
    public model: TModel;
}

export class ImportDatabaseResponse {
    public message: string;
    public didError: boolean;
    public errorMessage: string;
}

export class ImportedDatabase {
    public name: string;
    public tablesCount: number;
    public viewsCount: number;
    public details: string;
}

export class DatabaseTypeMap {
    public databaseType: string;
    public allowsLengthInDeclaration: boolean;
    public clrFullNameType: string;
}

export class DatabaseDetail {
    public name: string;
    public tables: TableDetail[];
    public views: ViewDetail[];
    public mappings: DatabaseTypeMap[];
}

export class TableDetail {
    public Schema: string;
    public Name: string;
    public ColumnsCount: number;
    public PrimaryKey: string;
    public Identity: string;
    public Details: string;
}

export class ViewDetail {
    public Schema: string;
    public Name: string;
    public ColumnsCount: number;
    public Identity: string;
    public Details: string;
}
